File: 919. Complete Binary Tree Inserter.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 */
var CBTInserter = function(root) {
    this.nodes = [];
    this.depth = 0;
    const stack = [root];
    while(stack.length) {
        const n = stack.shift();
        if (n) {
            this.nodes.push(n);
            if (this.nodes.length >= Math.pow(2, this.depth + 1) - 1) {
                this.depth++;
            }
            if (n.left) {
                stack.push(n.left);
            }
            if (n.right) {
                stack.push(n.right);
            }
        }
    }
};

/** 
 * @param {number} v
 * @return {number}
 */
CBTInserter.prototype.insert = function(v) {  //Brute force since we can always calculate the parent index based on the child node index with Math.floor(index / 2)
    const node = new TreeNode(v);
    this.nodes.push(node);
    let parentNode = this.nodes[Math.floor(this.nodes.length / 2) - 1];
    if (this.nodes.length % 2 === 0) {
        parentNode.left = node;
    } else {
        parentNode.right = node;
    }
    return parentNode.val;
};

/**
 * @return {TreeNode}
 */
CBTInserter.prototype.get_root = function() {
    return this.nodes[0];
};

File: Golang/Tree/919.Complete-binary-tree-inserter.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type CBTInserter struct {
	root *TreeNode
}

func Constructor(root *TreeNode) CBTInserter {
	return CBTInserter{
		root: root,
	}
}

func (this *CBTInserter) Insert(val int) int {
	stack := []*TreeNode{this.root}
	ret := -1
	for len(stack) > 0 {
		child := []*TreeNode{}
		for _, s := range stack {
			if s.Left != nil {
				child = append(child, s.Left)
			}
			if s.Right != nil {
				child = append(child, s.Right)
			}
		}
		// next layer is not full
		if len(stack)*2 > len(child) {
			if len(child)%2 == 0 {
				parentNode := stack[(len(child)+1)/2]
				ret = parentNode.Val
				parentNode.Left = &TreeNode{
					Val: val,
				}
			} else {
				parentNode := stack[len(child)/2]
				ret = parentNode.Val
				parentNode.Right = &TreeNode{
					Val: val,
				}
			}
			break
			// next layer is a full layer
		} else if len(child) == 0 {
			parentNode := stack[0]
			ret = parentNode.Val
			parentNode.Left = &TreeNode{
				Val: val,
			}
			break
		}
		stack = child
	}
	return ret
}

func (this *CBTInserter) Get_root() *TreeNode {
	return this.root
}
